File: packages/typescript/tests/auth.test.ts
import { beforeEach, describe, expect, it } from "vitest";
import { http, HttpResponse } from "msw";
import { AuthManager } from "../src/auth.ts";
import { BASE } from "./mocks/handlers.ts";
import { server } from "./mocks/server.ts";

describe("AuthManager digest + bearer", () => {
  let authCallCount: number;

  beforeEach(() => {
    authCallCount = 0;
  });

  it("authenticates with Digest auth-int and stores bearer token", async () => {
    server.use(
      http.post(`${BASE}/auth`, () => {
        authCallCount += 1;
        if (authCallCount === 1) {
          return new HttpResponse(
            JSON.stringify({
              status: { success: false, info: [{ code: 401, message: "Unauthorized" }] },
            }),
            {
              status: 401,
              headers: {
                "WWW-Authenticate":
                  'Digest realm="sonicwall", nonce="abc123", algorithm=SHA-256, qop="auth-int"',
              },
            }
          );
        }
        return HttpResponse.json({
          status: {
            success: true,
            info: [{ bearer_token: "test-bearer-token" }],
          },
        });
      })
    );

    const auth = new AuthManager(`${BASE}/`, "admin", "password");
    await auth.ensureAuthenticated({} as never);
    expect(auth.getBearerToken()).toBe("test-bearer-token");
    expect(auth.getSessionCookie()).toBeNull();
  });

  it("falls back to cookie auth when no auth-int challenge", async () => {
    server.use(
      http.post(`${BASE}/auth`, ({ request }) => {
        authCallCount += 1;
        if (authCallCount === 1) {
          return new HttpResponse(null, { status: 401 });
        }
        const authHeader = request.headers.get("authorization") ?? "";
        expect(authHeader.startsWith("Basic ")).toBe(true);
        return new HttpResponse(JSON.stringify({ status: { success: true, info: [] } }), {
          status: 200,
          headers: { "Set-Cookie": "smngsess=cookie-value; Path=/" },
        });
      })
    );

    const auth = new AuthManager(`${BASE}/`, "admin", "password");
    await auth.ensureAuthenticated({} as never);
    expect(auth.getSessionCookie()).toBe("cookie-value");
  });
});

File: packages/go/auth_test.go
package sonicwall

import (
	"context"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestAuthManagerDigestBearer(t *testing.T) {
	t.Parallel()
	callCount := 0
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/auth" {
			http.NotFound(w, r)
			return
		}
		callCount++
		if callCount == 1 {
			w.Header().Set(
				"WWW-Authenticate",
				`Digest realm="sonicwall", nonce="abc123", algorithm=SHA-256, qop="auth-int"`,
			)
			w.WriteHeader(http.StatusUnauthorized)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"status":{"success":true,"info":[{"bearer_token":"jwt-token"}]}}`))
	}))
	defer srv.Close()

	auth := newAuthManager(srv.URL, "admin", "password", srv.Client())
	if err := auth.ensureAuthenticated(context.Background()); err != nil {
		t.Fatalf("ensureAuthenticated: %v", err)
	}
	if got := auth.authorizationHeader(); got != "Bearer jwt-token" {
		t.Fatalf("authorizationHeader = %q", got)
	}
}

func TestPickAuthIntChallenge(t *testing.T) {
	t.Parallel()
	challenge := pickAuthIntChallenge([]string{
		`Digest realm="x", nonce="1", algorithm=MD5, qop="auth-int"`,
		`Digest realm="x", nonce="2", algorithm=SHA-256, qop="auth-int"`,
	})
	if challenge["algorithm"] != "SHA-256" {
		t.Fatalf("expected SHA-256, got %v", challenge["algorithm"])
	}
}
